toppra/dracula/three_segment_spline.py

```diff
@@ -0,0 +1,214 @@
+    P1 = (A0*m**2)/2 + (J0*m**3)/6 + P0 + m*V0
+    V1 = A0*m + (J0*m**2)/2 + V0
+    A1 = A0 + J0*m
+    J1 = -((A3*o**2 + A1*(3*n**2 + 6*n*o + 2*o**2) + 6*(P1 - P3 + (n + o)*V1))/(n*(n**2 + 3*n*o + 2*o**2)))
+    P2 = (A1*n**2)/2 + (J1*n**3)/6 + P1 + n*V1
+    V2 = A1*n + (J1*n**2)/2 + V1
+    A2 = A1 + J1*n
+    J2 = -((3*(A2*o**2 + 2*(P2 - P3 + o*V2)))/o**3)
+
+    if debugging:
+        V3test = V2 + A2*o + (1/2)*J2*(o**2)
+        A3test = A2 + J2*o
+        print('V3test',V3test)
+        print('A3test',A3test)
+        print('J0', J0, 'J1=',J1, 'J2', J2)
+        print('P1,V1,A1,J1 = ',P1,V1,A1,J1)
+        print('P2,V2,A2,J2 = ',P2,V2,A2,J2)
+
+    #Gather results
+    polynomial_coeff_m = np.stack((1/6*J0,1/2*A0,V0,P0))
+    polynomial_coeff_n = np.stack((1/6*J1,1/2*A1,V1,P1))
+    polynomial_coeff_o = np.stack((1/6*J2,1/2*A2,V2,P2))
+    x = np.array([m,m+n,m+n+o])
+    c = np.stack([polynomial_coeff_m,polynomial_coeff_n,polynomial_coeff_o],axis=1)
+    if debugging:
+        print('x=\n',x)
+        print('c=\n',c)
+    return x, c
+
+def ThreeSegmentRespline(cubic_spline, velocity_limits, acceleration_limits, jerk_limits):
+    shape = cubic_spline.c.shape
+    shape = list(shape)
+    shape[0] *= 3
+    shape = tuple(shape)
+    X_ret = np.empty(shape[0])
+    C_ret = np.empty(shape)
+    del shape
+
+    for i in range(cubic_spline.x.shape[0]-1):
+        X0 = cubic_spline.x[i]
+        P0 = cubic_spline(X0)
+        if i == 0:
+            # first segment
+            V0 = np.zeros(P0.shape)
+            A0 = np.zeros(P0.shape)
+        else:
+            V0 = cubic_spline(X0, 1)
+            A0 = cubic_spline(X0, 2)
+
+        X1 = cubic_spline.x[i+1]
+        P1 = cubic_spline(X1)
+        if i == cubic_spline.x.shape[0]-2:
+            # last segment
+            V1 = np.zeros(P1.shape)
+            A1 = np.zeros(P1.shape)
+        else:
+            V1 = cubic_spline(X1, 1)
+            A1 = cubic_spline(X1, 2)
+
+        X_new, C_new = ThreeSegmentSpline(P0, V0, A0, P1, V1, A1, X1 - X0)
+        X_ret[i:i+3] = X_new
+        C_ret[i:i+3] = C_new
+
+    delay = 0.0
+    for i in range(X_ret.shape[0]-1):
+        delta_X = X_ret[i+1] - X_ret[i]
+
+        scale = None
+        for joint_jerk, (joint_jerk_lower, joint_jerk_upper) in zip(C_ret[i,::,0] * 6, jerk_limits):
+            if joint_jerk < 0:
+                joint_scale = joint_jerk / joint_jerk_lower
+                if scale is None or scale < joint_scale:
+                    scale = joint_scale
+            elif joint_jerk > 0:
+                joint_scale = joint_jerk / joint_jerk_upper
+                if scale is None or scale < joint_scale:
+                    scale = joint_scale
+
+        if scale is None or scale < 1:
+            scale = 1
+        scale = scale ** (1/3)
+
+        C_ret[i,::,2] /= scale
+        C_ret[i,::,1] /= scale**2
+        C_ret[i,::,0] /= scale**3
```

Comment:
I couldn't wrap my head around the correct algebra for scaling the coefficients. I'll come back at this tomorrow